doc/wdread.1
.TH "wdread" "1" "15 August 2011"
.SH "NAME"
\fBwdread\fR \- read value from defaults database
.PP
.SH "SYNOPSIS"
.B wdread \fR[ \fI\-\-options \fR]
.PP
.SH "DESCRIPTION"
\fBwdread \fR\- read value from defaults database
.PP
.SH "OPTIONS"
.TP
.B \-\-help
\fRdisplay this help and exit
.TP
.B \-\-version
\fRoutput version information and exit
.TP
.B \-\-h
\fRlike \-\-help
.TP
.B \-\-v
\fRlike \-\-version
.PP
.SH "AUTHOR"
Window Maker was written by Alfredo K. Kojima <kojima@windowmaker.info>,
wdread is a part of Window Maker, was  cowardly remade from wdwrite.c;
by judas@hell on Jan 26 2001
.PP
This manpage was written by Rodolfo García Peñas (kix) for the
Debian GNU/Linux system (but may be used by others).


NEWS

NEWS for veteran Window Maker users
-----------------------------------

--- 0.94.0-crm

New Resizing functionality
--------------------------

You can now use the mouse wheel and modifier keys to resize windows.
MOD+Wheel will resize windows vertically and CTRL+Wheel will resize
windows horizontally. MOD+CTRL+Wheel will resize both at the same time.
The resize step, or increment can be set in WPrefs in the "Window Handling"
page, or in the config file with "ResizeIncrement". If this intrudes into
a specific application's functionality, you can disable it on an application-
by-application basis by setting the "Do not bind mouse clicks" attribute in
"Advanced Options".

New Maximize functionality
--------------------------

Maximus/Maximumize. A new tiled maximization, configured with a keyboard
shortcut. Using it will maximize the window to the greatest area such that it
will not overlap any other window of the same workspace. This can be configured
in WPrefs in the "Keyboard Shortcuts" page, or in the config file with
"MaximusKey".

Left/right maximization. Now wmaker supports maximizing a window to the
left or right half of the screen, making it occupy 50% of the area. This is
useful on widescreen displays where you want to bring up two windows side-by-
side. The shortcut keys for this can be configured in WPrefs in the "Keyboard
Shortcuts" page, or in the config file with "LHMaximizeKey" and "RHMaximizeKey"

History and AutoComplete in the run dialog
------------------------------------------

The "Run..." dialog now autocompletes executables in your path (triggered by
the TAB key) and records the history of past commands (last 500 by default).
It can also complete folder names if you start the complete with a "/"
To use it replace %a with %A in the "Run..." entry in your WMRootMenu file.
See commit 05720d97076ffc1569e5 for more details.

Automatic menu generator (wmgenmenu)
-----------------------------------

There is now a new utility to generate the Window Maker menu automatically,
called wmgenmenu. It searches from a list of pre-defined applications the
ones which exist in your $PATH and adds them to the corresponding submenus
("Internet", "Terminals", "Editors" etc) of your WM menu. It also supports
localization (currently English and German). You can use it like:

$ wmgenmenu > WMRootMenu-new
$ cp WMRootMenu-new $HOME/GNUstep/Defaults/WMRootMenu

and WM will automatically detect the new menu file and use it without
restarting wmaker (made possible by the 'inotify' mechanism)

Automatic detection of menu changes (via inotify)
-------------------------------------------------

Window Maker now uses the 'inotify' mechanism from the linux kernel
to automatically detect changes in the WMRootMenu file. So the
--no-polling option is now gone and Window Maker does not wake up
your CPU unnecessarily (0 wakeups when idle, instead of 4).
So if you edit the WMRootMenu file by hand (or by using 'wmgenmenu'),
there is no need to restart wmaker for the changes to take effect.

DockApp recognition
-------------------

In addition to applications with only Withdrawn windows, Window Maker
will now treat any application with its WM_CLASS res_class set as
"DockApp". This provides an easy workaround for toolkits like gtk+ that
do not allow creation of windows with the initial_state member of
XWMHints set to WithdrawnState.


--- 0.92.0

GNUstep Installation Directory
------------------------------

WPrefs is now installed in /usr/local/bin and /usr/local/share by default.
If you use GNUstep and want it to install in /usr/GNUstep/Applications,
you may specify --with-gnustepdir=/usr/GNUstep
If the GNUSTEP_LOCAL_ROOT environment variable is defined when configure is
executed, it will be used (and you don't need to use --with-gnustepdir)

Cached Pixmaps Directory
------------------------

The directory where Window Maker stores the cached application pixmaps for
its later use has changed from ~/GNUstep/.AppInfo/WindowMaker to
~/GNUstep/Library/WindowMaker/CachedPixmaps for better compatibility with
the GNUstep path structure.
Also WPrefs now stores internal data in ~/GNUstep/Library/WindowMaker/WPrefs
(it was ~/GNUstep/.AppInfo/WPrefs before)

X Input Methods support in WINGs
--------------------------------

Preliminary support for X Input Methods was added to textfield and text
widgets in WINGs. Input for text in other languages than English should
work now (except for kanji which will most likely not work, even though
it wasn't tested).


Disabling the switch panel
--------------------------

To disable the panel shown during Alt-tabbing, you may put the following in
~/GNUstep/Defaults/WindowMaker

SwitchPanelImages= None;



--- 0.91.0

Alt-Tab Window Switching
------------------------

You can change the appearance of the panel shown during Alt-Tab window switching
with the SwitchPanelImages option:

 (selected_icon_tile_image, background_image, width, height)

selected_icon_tile_image is the image used to highlight the currently selected
window icon. It must be 64x64 pixels.

background_image is the image used in the background of the panel. It must
be at least 64x80.

width and height are the width and size of the central part of the image.
When drawing the panel, the image will be split as:
                           |W |
			+--+--+--+
			|  |  |  |
			+--+--+--+ -
			|  |  |  | H
			+--+--+--+ -
			|  |  |  |
			+--+--+--+

The 4 corner images will be copied in their original sizes and the rest will
be scaled to the final panel size.

background_image, width and height are optional. If you leave them out,
a gray panel will be used. If your machine is not very fast, you may want 
to use it.


--- 0.90.0


NetWM / EWMH Support
--------------------

Support for the EWMH standard has been added. Applications from GNOME 2.x and
KDE 3.x should now interoperate better with Window Maker.

Support for the obsolete/legacy GNOME 1.x, KDE 1.x and OpenLook(!) hints was
wiped out for the sake of sanity.



util/wdread.c
/* wdread.c - read value from defaults database
 *
 *  WindowMaker window manager
 *
 *  Copyright (c) 1997-2003 Alfredo K. Kojima
 *  (cowardly remade from wdwrite.c; by judas@hell on Jan 26 2001)
 *
 *  This program is free software; you can redistribute it and/or modify
 *  it under the terms of the GNU General Public License as published by
 *  the Free Software Foundation; either version 2 of the License, or
 *  (at your option) any later version.
 *
 *  This program is distributed in the hope that it will be useful,
 *  but WITHOUT ANY WARRANTY; without even the implied warranty of
 *  MERCHANTABILITY or FITNESS FOR A PARTICULAR PURPOSE.  See the
 *  GNU General Public License for more details.
 *
 *  You should have received a copy of the GNU General Public License along
 *  with this program; if not, write to the Free Software Foundation, Inc.,
 *  51 Franklin Street, Fifth Floor, Boston, MA 02110-1301 USA.
 */

#ifdef __GLIBC__
#define _GNU_SOURCE		/* getopt_long */
#endif

/*
 * WindowMaker defaults DB reader
 */

#include <getopt.h>
#include <limits.h>
#include <stdio.h>
#include <stdlib.h>
#include <string.h>
#include <unistd.h>

#include <WINGs/WUtil.h>

#include "../src/wconfig.h"

extern char *__progname;

void print_help(int print_usage, int exitval)
{
	printf("Usage: %s [OPTIONS] <domain> <key>\n", __progname);
	if (print_usage) {
		puts("Read <key> from <domain>'s database");
		puts("");
		puts("  -h, --help              display this help message");
		puts("  -v, --version           output version information and exit");
	}
	exit(exitval);
}

int main(int argc, char **argv)
{
	char path[PATH_MAX];
	WMPropList *key, *value, *dict;
	int ch;

	struct option longopts[] = {
		{ "version",	no_argument,		NULL,			'v' },
		{ "help",	no_argument,		NULL,			'h' },
		{ NULL,		0,			NULL,			0 }
	};

	while ((ch = getopt_long(argc, argv, "hv", longopts, NULL)) != -1)
		switch(ch) {
			case 'v':
				printf("%s (Window Maker %s)\n", __progname, VERSION);
				return 0;
				/* NOTREACHED */
			case 'h':
				print_help(1, 0);
				/* NOTREACHED */
			case 0:
				break;
			default:
				print_help(0, 1);
				/* NOTREACHED */
		}

	argc -= optind;
	argv += optind;

	if (argc != 2)
		print_help(0, 1);

	key = WMCreatePLString(argv[1]);

	snprintf(path, sizeof(path), "%s", wdefaultspathfordomain(argv[0]));

	dict = WMReadPropListFromFile(path);
	if (dict == NULL)
		return 1;	/* bad domain */

	value = WMGetFromPLDictionary(dict, key);
	if (value == NULL)
		return 2;	/* bad key */

	printf("%s\n", WMGetPropListDescription(value, True));
	return 0;
}
